rustls-wolfcrypt-provider/src/aead/quic.rs

```diff
@@ -0,0 +1,1713 @@
+    key_len: AES_256_KEY_LEN,
+    hp_mask: generate_mask_aes,
+    id: HPAlgorithmID::Aes256,
+    init: init_hp_aes_cipher,
+};
+
+fn init_hp_aes_cipher(key: &[u8]) -> Result<Cipher, Error> {
+    let mut aes_cipher = AesCipher::default();
+    aes_cipher.set_key(key)?;
+    Ok(Cipher::Aes(aes_cipher))
+}
+
+fn generate_mask_aes(hp_cipher: &Cipher, sample: &[u8]) -> Result<[u8; MASK_LEN], Error> {
+    let aes_cipher = match hp_cipher {
+        Cipher::Aes(c) => c,
+        _ => return Err(Error::General("Invalid cipher type".into())),
+    };
+
+    let mut mask = mask_array!();
+    match aes_cipher.encrypt_sample(sample) {
+        Ok(output) => mask.copy_from_slice(&output[..5]),
+        Err(e) => return Err(e),
+    }
+    Ok(mask)
+}
+
+/// ChaCha20.
+pub static CHACHA20: HPAlgorithm = HPAlgorithm {
+    key_len: CHACHA_KEY_LEN,
+    init: init_hp_chacha20_cipher,
+    hp_mask: generate_mask_chacha20,
+    id: HPAlgorithmID::ChaCha20,
+};
+
+fn init_hp_chacha20_cipher(key: &[u8]) -> Result<Cipher, Error> {
+    let chacha_cipher = ChaChaCipher::new(None);
+    chacha_cipher.set_key(key)?;
+    Ok(Cipher::ChaCha20(chacha_cipher))
+}
+
+fn generate_mask_chacha20(hp_cipher: &Cipher, sample: &[u8]) -> Result<[u8; MASK_LEN], Error> {
+    let chacha20_cipher = match hp_cipher {
+        Cipher::ChaCha20(c) => c,
+        _ => return Err(Error::General("Invalid cipher type".into())),
+    };
+
+    let mut mask = mask_array!();
+    match chacha20_cipher.encrypt_sample(sample) {
+        Ok(output) => mask.copy_from_slice(&output[..5]),
+        Err(e) => return Err(e),
+    }
+    Ok(mask)
+}
+
+/// A key for generating QUIC Header Protection masks.
+pub struct HeaderProtectionKey {
+    hp_cipher: Cipher,
+    algorithm: &'static HPAlgorithm,
+}
+
+impl HeaderProtectionKey {
+    /// Create a new header protection key.
+    ///
+    /// `key_bytes` must be exactly `algorithm.key_len` bytes long.
+    pub fn new(key: Vec<u8>, algorithm: &'static HPAlgorithm) -> Result<Self, Error> {
+        if key.len() != algorithm.key_len {
+            return Err(Error::General("Invalid key length".into()));
+        }
+        Ok(Self {
+            hp_cipher: (algorithm.init)(&key)?,
+            algorithm,
+        })
+    }
+
+    fn header_protection(
+        &self,
+        sample: &[u8],
+        first: &mut u8,
+        packet_number: &mut [u8],
+        masked: bool,
+    ) -> Result<(), Error> {
+        // This implements "Header Protection Application" almost verbatim.
+        // <https://datatracker.ietf.org/doc/html/rfc9001#section-5.4.1>
+
+        if sample.len() != SAMPLE_LEN {
+            return Err(Error::General("Invalid sample length".into()));
+        }
+
+        let mask = (self.algorithm.hp_mask)(&self.hp_cipher, sample)?;
+
+        // The `unwrap()` will not panic because `new_mask` returns a
+        // non-empty result.
+        let (first_mask, pn_mask) = mask.split_first().unwrap();
```

Comment:
Let's avoid `unwrap`. I see rust has some ways to use `unwrap` without panicking. libs shouldn't take down an app. Maybe `unwrap_or_else` is appropriate?